src/serde/ser.rs

```diff
@@ -1,12 +1,43 @@
 use std::io;
 use std::io::Cursor;
+use std::io::ErrorKind;
+use std::io::Write;
 
 use super::write_atom::write_atom;
 use crate::allocator::{NodePtr, SExp};
 use crate::node::Node;
 
 const CONS_BOX_MARKER: u8 = 0xff;
 
+pub struct LimitedWriter<W: io::Write> {
+    inner: W,
+    limit: i64,
```

Comment:
```suggestion
    limit: usize,
```
This will propagate a few other simple (and annoying) changes.